### PR TITLE
(Just an) I18n framework

### PR DESCRIPTION
**Pull request by Lyonlancer5**
# Overview
Status: **Needs testing in other system setups**
Note: **A majority of the additions are related to changing the strings to translate-ables**

This pull request aims to provide opsu! with an I18n framework that would allow non-English speakers to load the game into their native language and would also provide with a bit of fixes into the current state of the game.

**About the Language Manager (internally called `LocaleManager`)**

The way the manager loads the languages is that they are directly read from the JAR file itself and stored in key-value pairs (e.g. `my.key=Value`). The options panel provides a way to easily change the current language **without having to reload the game** and to reload the language files from JAR if modified during run-time.

Also, the list of languages in the options grows as you add more language files into the JAR, therefore users can _add their own languages_ into the game without having to modify the source code.

# Changes:
- **Provides a reloadable "Language Manager" to translate the game**
- Update Gradle _wrapper_ (`2.3` -> `3.5`)
- Logging scheme now allows printing to both console _and_ file at the same time. ~~Defaults to verbose.~~
- glFlush() after Display.update() <- [This post about latency testing](https://blog.ppy.sh/2017-05-03/)
- ~~Added a way to load "unsupported" beatmaps (CtB only atm)~~
- `setVSync()` double-checking when changing FPS
- ~~Parallax in beatmap playfield~~
- FPS string is now colored depending on target framerate
- ~~Change GC memory threshold (from `150,000,000` bytes to `157,286,400` bytes)~~
- ~~(Eclipse) Remove redundant type modifiers~~

### Possible fixes to:
- #276 ([`SoundController.stopSound()`](https://github.com/itdelatrisu/opsu/commit/ef79b3c680db5788e5b0331f2747004580791199#diff-03e43f23230df6fb8aa84823786af1c6) was... [a bit of a hack](https://github.com/itdelatrisu/opsu/commit/ef79b3c680db5788e5b0331f2747004580791199#diff-a7fc85db85bbf56ca25e1a40eba9ef10).)

### Benefits
- Let non-English speakers play the game at their own language.
- ~~Threaded loading now loads opsu! slightly faster (although sometimes causes OpenAL to not load itself)~~
- GL flushing provides a ~5% framerate boost, however, the effect is dependent on the graphics card _and_ drivers. 
(`Test specs: Intel Pentium E5700@3.00GHz, 3GB DDR2, nVidia 9500 GT using Microsoft's HCL driver in Windows 8.1`)

### On a side note...
~~This was supposed to be only an experiments branch, but since things "work better," I also provided certain edits that would help in making the game "_faster_" (although "_more responsive_" and "_less laggy_" would be better)~~

🕐  :octocat: 💤

### CODE REVIEW

**Comment by Lyonlancer5** [comment]
Update #293 and pinging [Lyonlancer5/opsu#2](https://github.com/Lyonlancer5/opsu/issues/2) for reference.
~~Self-note: C#, ye, _definitely like Java_... /s, short functions in other languages need to be "longer" with Java... orz~~

**Comment by itdelatrisu** [comment]
Thanks so much!  Very much appreciated. :) The GL flushing looks interesting; I hadn't seen that post before.

First off, please separate this into multiple pull requests (one per feature) or this will take a long time to get merged... reviewing an 1800-line diff is too much.  From glancing at the code, I also see a lot of code style changes and one-line edits/additions that aren't explained. Please leave these out of this pull request.

From reading your comments, I'm confused about why you made some of the other changes:
* I don't see the point of extending the logging system. Also, it's not set to verbose by default so that users don't see as much garbage in their logs when filing issues.
* I don't see why someone would ever want to load beatmaps for non-implemented modes.
* There's no parallax in the beatmap playfield in osu! (at least for me).
* Please don't remove type modifiers. It decreases the readability of the code.
* Why did you change the GC threshold?
* I disagree with spawning so many threads on startup. This project targets lower-end machines, so it's not supposed to be a CPU hog.

**Comment by tpenguinltg** [comment]
`NullPointerException` at startup when trying to run it:

```
# Log-to-file : Thu May 18 23:24:55 EDT 2017
[23:24:55] [main/INFO] [opsu!]: Initializing opsu!
[23:24:55] [main/INFO] [opsu!]: Loaded translations
[23:24:55] [main/INFO] [opsu!]: Unable to determine Slick build number
[23:24:55] [main/INFO] [opsu!]: LWJGL Version: 2.9.3
[23:24:55] [main/INFO] [opsu!]: OriginalDisplayMode: 1366 x 768 x 24 @60Hz
[23:24:55] [main/INFO] [opsu!]: TargetDisplayMode: 800 x 600 x 0 @0Hz
[23:24:55] [main/INFO] [opsu!]: Starting display 800x600
[23:24:55] [main/INFO] [opsu!]: Use Java PNG Loader = true
[23:24:58] [main/ERROR] [opsu!]: null
java.lang.NullPointerException
	at itdelatrisu.opsu.translations.LocaleManager.translateKey(LocaleManager.java:278)
	at itdelatrisu.opsu.GameMod$Category.getName(GameMod.java:112)
	at itdelatrisu.opsu.GameMod.init(GameMod.java:193)
	at itdelatrisu.opsu.Utils.init(Utils.java:158)
	at itdelatrisu.opsu.states.Splash.init(Splash.java:108)
	at org.newdawn.slick.state.StateBasedGame.init(StateBasedGame.java:171)
	at org.newdawn.slick.AppGameContainer.setup(AppGameContainer.java:393)
	at itdelatrisu.opsu.Container.runSetup(Container.java:114)
	at itdelatrisu.opsu.Container.start(Container.java:79)
	at itdelatrisu.opsu.Opsu.main(Opsu.java:211)
```

Arch Linux, Java 1.8.0_102, gradle 3.4.1

**Comment by Lyonlancer5** [comment]
err... I may have committed an error and would need to clear this out first.
I'm closing this first so that I can fix this

//in hindsight, maybe it was better to use GitHub's client rather than Eclipse's since I'm already used to Github's client

**Comment by Lyonlancer5** [comment]
I asked help from a close friend  (@Leidi-Kana) to help resolve the accidental 15,000 changes commit. Apparently, changing the file "dates" causes git to treat them as different files.

He said that the easiest way to fix it was to just "revert the previous commits"
And that I did. Sorry for this _huge_ mess. I'll make better PRs next time.

-----

Now then, about a few things:

> `NullPointerException` at startup when trying to run it

This is caused by LocaleManager not being able to load the language files from disk, and the solution to that is to search _both the JAR and the current directory_ for languages and hard-code the default English translation into the class. (adds +200 lines of code)

> I don't see the point of extending the logging system.

This logging implementation was designed to provide thread information to the logs. This, in some cases, helps determine what threads are doing at the time of logging. I'll remove it if it isn't really required.

> There's no parallax in the beatmap playfield in osu! (at least for me).

I intended to replicate a feature from [osu!lazer](https://github.com/ppy/osu) which has parallax in the playfield. Removed, should have posted it in `experiments1`
Removed, should have committed in `experiments1`

> I don't see why someone would ever want to load beatmaps for non-implemented modes

~~cursor-dance and~~ since CtB/Taiko maps are technically "mapped in Standard"
**Removed**, should have committed in `experiments1`

> GC threshold

I was testing for the "minimum system requirements", especially in memory.
**Reverted**.

> I disagree with spawning so many threads on startup. This project targets **lower-end machines**, so it's not supposed to be a CPU hog.

Can we make this a toggle-able option instead? (for slightly more powerful PCs)
This is related to the GC threshold testing.
**Reverted**.

**Comment by yugecin** [comment]
Slick already does a flush if I'm not mistaken: https://github.com/itdelatrisu/opsu/blob/8aeb4ed2e06d0f05ffdd17e216406a49e2f8ce5a/src/org/newdawn/slick/GameContainer.java#L738

@itdelatrisu maybe a contributing guide (with rules like keeping type modifiers, and using tabs instead of spaces, whitespaces, line length etc ..) would be useful for uniformity & PRs.

**Comment by itdelatrisu** [comment]
@Lyonlancer5: Thanks for figuring that out. Sorry if I was unclear earlier -- I'm happy to discuss the other changes with you, but it makes more sense to have them in a different pull request.  If it's not too much work for you to really isolate the i18n changes for now, that'd be appreciated (since it's such a huge change).  I'll make a quick pass over the diff soon, though (within a day, hopefully sooner).

@yugecin: Thanks for pointing that out, I hadn't checked yet.  Also, I'll try to put up a short contributing guide soon, that's a good idea.  Generally I haven't really cared (I just reformat the new code myself afterwards), but I suppose it'll give a good starting point.

File: src/itdelatrisu/opsu/io/OsuWriter.java [hunk 1]
```diff
@@ -52,7 +52,11 @@ public OsuWriter(File file) throws FileNotFoundException {
 	 * @param dest the output stream to write to
 	 */
 	public OsuWriter(OutputStream dest) {
-		this.writer = new DataOutputStream(new BufferedOutputStream(dest));
+		if(dest instanceof BufferedOutputStream){
```
Comment (itdelatrisu) [inline src/itdelatrisu/opsu/io/OsuWriter.java]: ? (why is this needed, and if it is, why not a separate constructor?)

File: src/itdelatrisu/opsu/options/Options.java [hunk 2]
```diff
@@ -196,7 +196,7 @@ private static File getXDGBaseDir(String env, String fallback) {
 	}
 
 	/**
-	 * Returns the osu! installation directory.
+	 * Returns the osu! installation directory (currently supports Windows only)
```
Comment (itdelatrisu) [inline src/itdelatrisu/opsu/options/Options.java]: unnecessary (osu! only runs on Windows anyway...)

File: src/itdelatrisu/opsu/options/Options.java [hunk 3]
```diff
@@ -224,7 +224,7 @@ private static File getOsuInstallationDirectory() {
 	}
 
 	/** Game options. */
-	public enum GameOption {
+	public static enum GameOption {
```
Comment (itdelatrisu) [inline src/itdelatrisu/opsu/options/Options.java]: ?

File: src/itdelatrisu/opsu/options/Options.java [hunk 4]
```diff
@@ -729,7 +755,8 @@ public String getValueString() {
 		 * Returns the option name.
 		 * @return the name string
 		 */
-		public String getName() { return name; }
+		//public String getName() { return name; }
```
Comment (itdelatrisu) [inline src/itdelatrisu/opsu/options/Options.java]: delete line

File: src/itdelatrisu/opsu/options/Options.java [hunk 5]
```diff
@@ -741,7 +768,8 @@ public String getValueString() {
 		 * Returns the option description.
 		 * @return the description string
 		 */
-		public String getDescription() { return description; }
+		//public String getDescription() { return description; }
```
Comment (itdelatrisu) [inline src/itdelatrisu/opsu/options/Options.java]: delete line

File: src/itdelatrisu/opsu/options/OptionsOverlay.java [hunk 6]
```diff
@@ -238,6 +238,7 @@
 	private GameContainer container;
 	private Input input;
 	private int containerWidth, containerHeight;
+
```
Comment (itdelatrisu) [inline src/itdelatrisu/opsu/options/OptionsOverlay.java]: ?

File: src/itdelatrisu/opsu/replay/ReplayImporter.java [hunk 7]
```diff
@@ -112,8 +112,8 @@ public boolean accept(File dir, String name) {
 		files = null;
 
 		if (importCount > 0) {
-			String text = String.format("Imported %d replay%s.", importCount, importCount == 1 ? "" : "s");
-			UI.getNotificationManager().sendNotification(text);
+			UI.getNotificationManager().sendNotification("ui.notifications.replay.imported", 
+					importCount /*, importCount == 1 ? "" : "s"*/);
```
Comment (itdelatrisu) [inline src/itdelatrisu/opsu/replay/ReplayImporter.java]: (same comment as above about plurals)

File: src/itdelatrisu/opsu/skins/Skin.java [hunk 8]
```diff
@@ -373,4 +373,4 @@ public Skin(File dir) {
 	 * Returns the amount of overlap between the combo font sprites.
 	 */
 	public int getComboFontOverlap() { return comboOverlap; }
-}
+}
```
Comment (itdelatrisu) [inline src/itdelatrisu/opsu/skins/Skin.java]: ?

File: src/itdelatrisu/opsu/skins/SkinLoader.java [hunk 9]
```diff
@@ -296,4 +296,4 @@ private static boolean isValidLine(String line) {
 		tokens[1] = line.substring(index + 1).trim();
 		return tokens;
 	}
-}
+}
```
Comment (itdelatrisu) [inline src/itdelatrisu/opsu/skins/SkinLoader.java]: ?

File: src/itdelatrisu/opsu/skins/SkinUnpacker.java [hunk 10]
```diff
@@ -77,8 +77,8 @@ public boolean accept(File dir, String name) {
 		files = null;
 
 		if (!dirs.isEmpty()) {
-			String text = String.format("Imported %d new skin%s.", dirs.size(), dirs.size() == 1 ? "" : "s");
-			UI.getNotificationManager().sendNotification(text);
+			//String text = String.format("Imported %d new skin%s.", dirs.size(), dirs.size() == 1 ? "" : "s");
+			UI.getNotificationManager().sendNotification("Imported %d new skin%s.", dirs.size());
```
Comment (itdelatrisu) [inline src/itdelatrisu/opsu/skins/SkinUnpacker.java]: (same comment above about plurals)

File: src/itdelatrisu/opsu/states/ButtonMenu.java [hunk 11]
```diff
@@ -59,13 +60,13 @@
  */
 public class ButtonMenu extends BasicGameState {
 	/** Menu states. */
-	public enum MenuState {
+	public static enum MenuState {
```
Comment (itdelatrisu) [inline src/itdelatrisu/opsu/states/ButtonMenu.java]: ?

File: src/itdelatrisu/opsu/states/Game.java [hunk 12]
```diff
@@ -393,7 +393,16 @@ else if (deathTime > -1)  // "Easy" mod: health bar increasing
 			Graphics.setCurrent(gOffscreen);
 		}
 
-		// background
+		/* background
+		 * 
+		 * Layers:
```
Comment (itdelatrisu) [inline src/itdelatrisu/opsu/states/Game.java]: This can be a 1-line comment
```
// Layers (top to bottom): hit objects, storyboard, background image/video
```

File: src/itdelatrisu/opsu/states/GamePauseMenu.java [hunk 13]
```diff
@@ -130,6 +130,9 @@ else if (key == Options.getGameKeyRight())
 				MusicController.playAt(MusicController.getBeatmap().previewTime, true);
 				if (UI.getCursor().isBeatmapSkinned())
 					UI.getCursor().reset();
+
+				//Workaround for not-stopping sfx
```
Comment (itdelatrisu) [inline src/itdelatrisu/opsu/states/GamePauseMenu.java]: Wouldn't call it a "workaround", there's not really another way to do this...

File: src/itdelatrisu/opsu/states/GamePauseMenu.java [hunk 14]
```diff
@@ -170,8 +173,11 @@ public void mousePressed(int button, int x, int y) {
 		} else if (backButton.contains(x, y)) {
 			SoundController.playSound(SoundEffect.MENUBACK);
 			((SongMenu) game.getState(Opsu.STATE_SONGMENU)).resetGameDataOnLoad();
-			if (loseState)
+			if (loseState){
+				//Workaround for not-stopping sfx
```
Comment (itdelatrisu) [inline src/itdelatrisu/opsu/states/GamePauseMenu.java]: (same as above)

File: src/itdelatrisu/opsu/states/GameRanking.java [hunk 15]
```diff
@@ -267,6 +270,9 @@ private void returnToSongMenu() {
 		songMenu.resetGameDataOnLoad();
 		if (UI.getCursor().isBeatmapSkinned())
 			UI.getCursor().reset();
+		
+		//Workaround for not-stopping sfx
+		SoundController.stopSound(SoundEffect.APPLAUSE);
```
Comment (itdelatrisu) [inline src/itdelatrisu/opsu/states/GameRanking.java]: (same as above)

File: src/itdelatrisu/opsu/audio/SoundController.java [hunk 16]
```diff
@@ -311,6 +311,15 @@ private static void playClip(MultiClip clip, float volume, LineListener listener
 			}
 		}
 	}
+	
+	//TODO Workaround for applause/failsound not stopping
+	/**
+	 * Stops playing a sound, if it is currently active
+	 * @param s The sound effect
+	 */
+	public static void stopSound(SoundComponent s) {
```
Comment (itdelatrisu) [inline src/itdelatrisu/opsu/audio/SoundController.java]: For usage: You can just call this method for applause/fail on the `leave()` event in game states where they're played, instead of the places you're currently calling this.

File: src/itdelatrisu/opsu/states/SongMenu.java [hunk 17]
```diff
@@ -935,10 +935,9 @@ else if (reloadThread.isFinished()) {
 				hoverOffset.setTime(0);
 			}
 			return;
-		} else {  // not hovered
-			hoverOffset.setTime(0);
-			hoverIndex = null;
-		}
+		} // not hovered
```
Comment (itdelatrisu) [inline src/itdelatrisu/opsu/states/SongMenu.java]: ...?

File: src/itdelatrisu/opsu/states/SongMenu.java [hunk 18]
```diff
@@ -1872,8 +1871,9 @@ private void doStateActionOnLoad(MenuState menuState, BeatmapSetNode node, Score
 				scores[i].loadGlyphs();
 			}
 			return scores;
-		} else
-			return null;  // incorrect map
+		}
```
Comment (itdelatrisu) [inline src/itdelatrisu/opsu/states/SongMenu.java]: don't change code style

File: src/itdelatrisu/opsu/states/Splash.java [hunk 19]
```diff
@@ -121,13 +119,13 @@ public void update(GameContainer container, StateBasedGame game, int delta)
 			throws SlickException {
 		if (!init) {
 			init = true;
-
+			
 			// resources already loaded (from application restart)
 			if (BeatmapSetList.get() != null) {
 				if (newSkin || watchServiceChange) {  // need to reload resources
-					thread = new Thread() {
-						@Override
```
Comment (itdelatrisu) [inline src/itdelatrisu/opsu/states/Splash.java]: don't delete `@Override` annotations

File: src/itdelatrisu/opsu/states/Splash.java [hunk 20]
```diff
@@ -150,31 +147,26 @@ public void run() {
 
 			// load all resources in a new thread
 			else {
-				thread = new Thread() {
-					@Override
```
Comment (itdelatrisu) [inline src/itdelatrisu/opsu/states/Splash.java]: same as above

File: src/itdelatrisu/opsu/states/Splash.java [hunk 21]
```diff
@@ -150,31 +147,26 @@ public void run() {
...
+				
+				thread = new Thread(){
 					public void run() {
-						File beatmapDir = Options.getBeatmapDir();
```
Comment (itdelatrisu) [inline src/itdelatrisu/opsu/states/Splash.java]: why...?

File: src/itdelatrisu/opsu/states/Splash.java [hunk 22]
```diff
@@ -140,7 +138,6 @@ public void run() {
 							Utils.gc(true);
 
 							finished = true;
-							thread = null;
```
Comment (itdelatrisu) [inline src/itdelatrisu/opsu/states/Splash.java]: ?

File: src/itdelatrisu/opsu/states/Splash.java [hunk 23]
```diff
@@ -150,31 +147,26 @@ public void run() {
...
+						// import skins
+						SkinUnpacker.unpackAllFiles(Options.getImportDir(), Options.getSkinRootDir());
 
 						// load sounds
 						SoundController.init();
 
 						Utils.gc(true);
-
 						finished = true;
-						thread = null;
```
Comment (itdelatrisu) [inline src/itdelatrisu/opsu/states/Splash.java]: ?

File: src/itdelatrisu/opsu/translations/LocaleManager.java [hunk 24]
```diff
@@ -0,0 +1,608 @@
...
+
+import org.newdawn.slick.util.Log;
+
+import craterstudio.io.Streams;
+
+/**
+ * Provides translation capabilities to opsu!
+ * @author Lyonlancer5
+ */
+public class LocaleManager {
```
Comment (itdelatrisu) [inline src/itdelatrisu/opsu/translations/LocaleManager.java]: Fix the spacing in this class (use tabs)

File: src/itdelatrisu/opsu/translations/LocaleManager.java [hunk 25]
```diff
@@ -0,0 +1,608 @@
...
+            reader = new BufferedReader(new InputStreamReader(stream, "UTF-8"));
+            String line;
+            while ((line = reader.readLine()) != null ) {
+                if(!line.isEmpty() && line.charAt(0) != 35){
+                	String[] values = line.split("=", 2);
+                	this.keys.put(values[0], NUMERIC_VARIABLE_PATTERN.matcher(values[1]).replaceAll("%$1s"));
+                }
+            }
+        }
+        catch (IOException var7)
```
Comment (itdelatrisu) [inline src/itdelatrisu/opsu/translations/LocaleManager.java]: `var7`?

File: src/itdelatrisu/opsu/translations/LocaleManager.java [hunk 26]
```diff
@@ -0,0 +1,608 @@
...
+import java.util.HashMap;
+import java.util.IllegalFormatException;
+import java.util.Map;
+import java.util.jar.JarEntry;
+import java.util.jar.JarFile;
+import java.util.regex.Pattern;
+
+import org.newdawn.slick.util.Log;
+
+import craterstudio.io.Streams;
```
Comment (itdelatrisu) [inline src/itdelatrisu/opsu/translations/LocaleManager.java]: Don't use this, I'd rather not use obscure/untested third-party libraries unless absolutely necessary (e.g. for video).

File: src/itdelatrisu/opsu/ui/BackButton.java [hunk 27]
```diff
@@ -44,10 +45,12 @@
 
 	/** Target duration, in ms, of the button animations. */
 	private static final int ANIMATION_TIME = 500;
+	
+	private final GameContainer ownerObject;
```
Comment (itdelatrisu) [inline src/itdelatrisu/opsu/ui/BackButton.java]: Call this `container`.

File: src/itdelatrisu/opsu/ui/NotificationManager.java [hunk 28]
```diff
@@ -50,7 +51,10 @@
 		public void click();
 	}
 
-	/** Notification. */
+	/** 
+	 * The default bubble notification,
```
Comment (itdelatrisu) [inline src/itdelatrisu/opsu/ui/NotificationManager.java]: Not sure why you're changing these comments. Unnecessary "default" (no others) and they don't necessarily appear in the lower-right of the screen (if stacked).

File: src/itdelatrisu/opsu/ui/NotificationManager.java [hunk 29]
```diff
@@ -334,25 +338,37 @@ public void update(int delta) {
...
 	 * Submits a bubble notification for drawing.
-	 * @param s the notification string
+	 * This method uses the default {@link Color#white} when drawing the notification borders
+	 * and does not have any associated {@link NotificationListener}s associated with it.
+	 * @param msg the notification string
+	 * @param args The formatting for the notification string, if any
 	 */
-	public void sendNotification(String s) { sendNotification(s, Color.white); }
-
+	public void sendNotification(String msg, Object... args) {
```
Comment (itdelatrisu) [inline src/itdelatrisu/opsu/ui/NotificationManager.java]: Note: might want to overload all of these methods without the varargs. (not sure if possible, can check later)

File: src/itdelatrisu/opsu/ui/UI.java [hunk 30]
```diff
@@ -103,7 +103,7 @@ public static void init(GameContainer container, StateBasedGame game) {
 		// notification manager
 		notificationManager = new NotificationManager(container);
 	}
-
+
```
Comment (itdelatrisu) [inline src/itdelatrisu/opsu/ui/UI.java]: ?

File: src/itdelatrisu/opsu/util/ListSet.java [hunk 31]
```diff
@@ -0,0 +1,178 @@
...
+import java.util.ListIterator;
+import java.util.Set;
+
+/**
+ * A <code>List</code> implementation that does not
+ * allow duplicate elements.
+ * 
+ * @author Lyonlancer5
+ */
+public class ListSet<E> implements List<E>, Set<E> {
```
Comment (itdelatrisu) [inline src/itdelatrisu/opsu/util/ListSet.java]: You'll have to convince me why this class is necessary...

File: src/org/newdawn/slick/GameContainer.java [hunk 32]
```diff
@@ -843,7 +843,11 @@ public void setTargetFrameRate(int fps) {
 	 */
 	public void setVSync(boolean vsync) {
 		this.vsync = vsync;
-		Display.setVSyncEnabled(vsync);
+		try{
+			Display.setVSyncEnabled(vsync);
+		} catch (Exception e) {
+			Log.error("GL error", e);
```
Comment (itdelatrisu) [inline src/org/newdawn/slick/GameContainer.java]: Why are you suppressing these errors?

File: src/org/newdawn/slick/openal/SoundStore.java [hunk 33]
```diff
@@ -52,7 +52,6 @@
  * @author Kevin Glass
  * @author Rockstar setVolume cleanup 
  */
-@SuppressWarnings({"rawtypes", "unchecked", "unused"})
```
Comment (itdelatrisu) [inline src/org/newdawn/slick/openal/SoundStore.java]: Don't edit the files I'm overriding from third-party libraries -- the more changes, the harder it is to diff against the originals.

File: src/itdelatrisu/opsu/options/Options.java (hunk 2 above)
Comment (Lyonlancer5) [inline src/itdelatrisu/opsu/options/Options.java]: > osu! only runs on Windows anyway...

Wine and PlayOnLinux allow osu! to be _run_ from Linux machines, although this is considered a "hack" because it was never intended to be _played_ on Linux. (Although the new client allows it to be compiled and run on Linux, but its status is still under Alpha and its really incomplete as of this PR.)

And I think what you meant to say was:
> osu! only runs **natively** on Windows anyway...

**Removed additional comment** as per request

File: src/itdelatrisu/opsu/translations/LocaleManager.java (hunk 24 above)
Comment (Lyonlancer5) [inline src/itdelatrisu/opsu/translations/LocaleManager.java]: Something like `Ctrl + Shift + F` in Eclipse?

**Fixed formatting** as per request/code style conformity

File: src/itdelatrisu/opsu/GameData.java [hunk 34]
```diff
@@ -756,8 +756,10 @@ else if (health.getHealth() >= 25f)
 			// combo burst
 			if (comboBurstIndex != -1 && comboBurstAlpha > 0f) {
 				Image comboBurst = comboBurstImages[comboBurstIndex];
+				//TODO flip image if drawn on the right side of the screen
+				//if (comboBurstX <= width - comboBurst.getWidth()) comboBurst = comboBurst.getFlippedCopy(true, false);
```
Comment (Lyonlancer5) [inline src/itdelatrisu/opsu/GameData.java]: [The image flips **after** its done moving to the side.](https://gfycat.com/BossyLiveGharial)
**Removed** for "cleaner" code

File: src/itdelatrisu/opsu/beatmap/OszUnpacker.java [hunk 35]
```diff
@@ -83,8 +83,7 @@ public boolean accept(File dir, String name) {
 		files = null;
 
 		if (!dirs.isEmpty()) {
-			String text = String.format("Imported %d new beatmap pack%s.", dirs.size(), dirs.size() == 1 ? "" : "s");
-			UI.getNotificationManager().sendNotification(text);
+			UI.getNotificationManager().sendNotification("ui.notifications.beatmaps.imported", dirs.size());
```
Comment (Lyonlancer5) [inline src/itdelatrisu/opsu/beatmap/OszUnpacker.java]: Added the "hack" for plurals since the formatting implementation allow us to do so.

> By default, `String.format()` ignores the extra arguments if the string does not contain enough formatting tags.

> **But**, if `String.format()` has more formatting tags than the arguments, it will throw a `MissingFormatArgumentException`

> This also applies if the formatting tag is of the incorrect type, but throws a different exception. (e.g. using `%d` instead of `%s` on a `String` format argument)

File: src/itdelatrisu/opsu/audio/SoundController.java (hunk 16 above)
Comment (Lyonlancer5) [inline src/itdelatrisu/opsu/audio/SoundController.java]: **Fixed**, now only called on `leave()` in game states.

**Comment by Lyonlancer5** [comment]
Will be closing this PR to redo the whole I18n framework without the extra features and separating them into their own specific request instead

(Sorry about the mess)

**Comment by itdelatrisu** [comment]
Thanks, sorry for the trouble.
